quickwit-metastore/src/metastore/postgresql_metastore.rs

```diff
@@ -732,6 +734,187 @@ impl Metastore for PostgresqlMetastore {
+                "#,
+            )
+            .bind(index_id)
+            .fetch_optional(tx)
+            .await
+            .map_err(|error| MetastoreError::DbError {
+                message: error.to_string(),
+            })?
+            .map(|row| {
+                let opstamp: i64 = row.get(0);
+                opstamp as u64
+            })
+            .unwrap_or(0);
+            Ok(max_opstamp)
+        })
+    }
+
+    /// Creates delete task.
+    #[instrument(skip(self),fields(index_id=delete_query.index_id.as_str()))]
+    async fn create_delete_task(&self, delete_query: DeleteQuery) -> MetastoreResult<DeleteTask> {
+        run_with_tx!(self.connection_pool, tx, {
+            // Serialize the delete query to fit the database model.
+            let delete_query_json = serde_json::to_string(&delete_query).map_err(|err| {
+                MetastoreError::InternalError {
+                    message: "Failed to serialize delete query.".to_string(),
+                    cause: err.to_string(),
+                }
+            })?;
+            // Create delete task.
+            let (create_timestamp, opstamp): (sqlx::types::time::PrimitiveDateTime, i64) =
+                sqlx::query(
+                    "INSERT INTO delete_tasks (index_id, delete_query_json) VALUES ($1, $2) \
+                     RETURNING create_timestamp, opstamp",
+                )
+                .bind(&delete_query.index_id)
+                .bind(&delete_query_json)
+                .fetch_one(tx)
+                .await
+                .map(|row| (row.get(0), row.get(1)))
+                .map_err(|err| convert_sqlx_err(&delete_query.index_id, err))?;
+            Ok(DeleteTask {
+                create_timestamp: create_timestamp.assume_utc().unix_timestamp(),
+                opstamp: opstamp as u64,
+                delete_query,
+            })
+        })
+    }
+
+    /// Deletes delete tasks of a given index.
+    #[instrument(skip(self))]
+    async fn delete_delete_tasks(&self, index_id: &str) -> MetastoreResult<()> {
+        run_with_tx!(self.connection_pool, tx, {
+            sqlx::query("DELETE FROM delete_tasks WHERE index_id = $1")
+                .bind(index_id)
+                .execute(tx)
+                .await?;
+            Ok(())
+        })
+    }
+
+    /// Update splits delete opstamps.
+    #[instrument(skip(self))]
+    async fn update_splits_delete_opstamp<'a>(
+        &self,
+        index_id: &str,
+        split_ids: &[&'a str],
+        delete_opstamp: u64,
+    ) -> MetastoreResult<()> {
+        if split_ids.is_empty() {
+            return Ok(());
+        }
+        run_with_tx!(self.connection_pool, tx, {
+            let updated_split_ids: Vec<String> = sqlx::query(
+                r#"
+                UPDATE splits
+                SET delete_opstamp = $1
+                WHERE
+                    index_id = $2
+                    AND split_id = ANY($3)
+                    AND split_state = $4
```

Comment:
```suggestion
                    AND split_state = $4
```

Why should we check for the state?

Reply:
You're right yes, I should have removed this condition, when I first coded it I wanted to make sure to do that only on published splits because that's what we should do. But the responsibility of this is on the caller anyway. Ideally we would want to only update these stamps only on mature splits but this condition can only be check outside the metastore.

Reply:
Done.